proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,483 @@
+  instance Interpolate RawHtml Html where
+    interpolate = toBuilder . Html . unRawHtml
+  instance {-# OVERLAPPABLE #-} Show a => Interpolate a Html where
+    interpolate = interpolate . show
+
+And gain access to safe string interpolation with HTML escaping by default:
+
+::
+
+  let title = "Why is 1 > 0?" :: Text
+  let body = "<p>Hello world</p>" :: Text
+
+  s"<h1>${title}</h1>${RawHtml body}"
+    == Html "<h1>Why is 1 &gt; 0?</h1><p>Hello world</p>"
+
+Custom interpolatable type: BigDecimal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Imagine a library implements a new ``BigDecimal`` type:
+
+::
+
+  data BigDecimal = BigDecimal Integer Int
+
+  renderBigDecimal :: BigDecimal -> String
+  renderBigDecimal (BigDecimal digits scale) =
+    let (int, frac) = splitAt scale (show digits)
+     in int <> "." <> frac
+
+That library could define the following instances:
+
+::
+
+  instance Interpolate BigDecimal String where
+    interpolate = interpolate . renderBigDecimal
+  instance Interpolate BigDecimal Text where
+    interpolate = interpolate . Text.pack . renderBigDecimal
+
+And be able to use it in interpolated strings:
+
+::
+
+  let n = BigDecimal 123456 3
+  s"123456 / 10^3 = ${n}" == "123456 / 10^3 = 123.456"
+
+Effect and Interactions
+-----------------------
+
+When ``-XOverloadedStrings`` is enabled, string interpolation can be used for any ``Buildable`` type. Otherwise, it will only ever build Strings.
+
+Interpolation is also supported with ``-XMultilineStrings``, as described in "Proposed Change Specification".
+
+Costs and Drawbacks
+-------------------
+
+Development should be low-effort, maintenance should be low-effort. Learnability for novice users will go up, since novice users probably expect string interpolation to be available, and might be frustrated at the lack of support currently.
+
+The major drawback of this approach is the typeclass instances problem:
+
+#. A new interpolator type (e.g. ``SqlQuery``) needs to define ``Builder`` and ``Interpolate`` for all known interpolatable types
+#. A new interpolatable type (e.g. ``BigDecimal``) needs to define ``Interpolate`` for all known interpolator types
+
+This is worse than ``IsString`` or ``Show`` due to the multi-param ``Interpolate`` type class. This makes ``Interpolate`` much more susceptible to orphan instances.
+
+One minor drawback is that whitespace is now important with this syntax, with ``s"foo"`` semantically different from ``s "foo"``. While there's precedent for this (Template Haskell splices make ``$(...)`` different from ``$ (...)``), this is the first instance where whitespace matters for an alphanumeric identifier. But IMO this isn't that big of a deal:
+
+#. It's unlikely for someone to be naming a function as ``s`` in the first place
+#. Prefixing string literals like ``s"..."`` is common in other languages: Python, Scala, Javascript/Typescript, etc.
+#. Easily mitigatable: just add a space, which improves readability anyway
+
+Alternatives
```

Comment:
I think it's worth briefly mentioning other possible designs, like the ones found in other interpolation libraries.

Reply:
I added a few comments, but the bulk of the design work found in other interpolation libraries is how to indicate placeholders and put the values back into the placeholders. With this work being done in the compiler, we don't need to do anything fancy here outside of `<>`, so I'm not sure how much we could bring in from other libraries here